mlos_bench/mlos_bench/schedulers/base_scheduler.py

```diff
@@ -74,14 +76,69 @@ def __init__(self, *,
 
         self._do_teardown = bool(config.get("teardown", True))
 
-        self.experiment: Optional[Storage.Experiment] = None
-        self.environment = environment
-        self.optimizer = optimizer
-        self.storage = storage
+        self._experiment: Optional[Storage.Experiment] = None
+        self._trial_runners = trial_runners
+        assert self._trial_runners, "At least one TrialRunner is required"
+        self._optimizer = optimizer
+        self._storage = storage
         self._root_env_config = root_env_config
+        self._current_trial_runner_idx = 0
 
         _LOG.debug("Scheduler instantiated: %s :: %s", self, config)
 
+    @property
+    def experiment(self) -> Optional[Storage.Experiment]:
+        """Gets the Experiment Storage."""
+        return self._experiment
+
+    @property
+    def root_environment(self) -> Environment:
+        """
+        Gets the root (prototypical) Environment from the first TrialRunner.
+
+        Note: This all TrialRunners have the same Environment config and are made
+        unique by their use of the unique trial_runner_id assigned to each
+        TrialRunner's Environment's global_config.
+        """
+        return self._trial_runners[0].environment
+
+    @property
+    def trial_runners(self) -> List[TrialRunner]:
+        """Gets the list of Trial Runners."""
+        return self._trial_runners
+
+    @property
+    def environments(self) -> Iterable[Environment]:
+        """Gets the Environment from the TrialRunners."""
+        return (trial_runner.environment for trial_runner in self._trial_runners)
+
+    @property
+    def optimizer(self) -> Optimizer:
+        """Gets the Optimizer."""
+        return self._optimizer
+
+    @property
+    def storage(self) -> Storage:
+        """Gets the Storage."""
+        return self._storage
+
+    def get_trial_runner(self, trial: Storage.Trial) -> TrialRunner:
+        """
+        Gets the TrialRunner associated with the given Trial.
+
+        Parameters
+        ----------
+        trial : Storage.Trial
+            The trial to get the associated TrialRunner for.
+
+        Returns
+        -------
+        TrialRunner
+        """
+        if trial.trial_runner_id is None:
+            raise ValueError(f"Trial {trial} has no trial_runner_id")
```

Comment:
-[ ] Add fallback support for assigning a Trial to a TrialRunner if one is missing.